Apt repo fixes #16370 (see #4072 as well for devel fix)

alikins opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

packaging/os/apt_repository.py
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.1.1.0 (stable-2.1 4d4bbcbb33) last updated 2016/06/20 18:17:18 (GMT +000)
  lib/ansible/modules/core: (detached HEAD b746f047ed) last updated 2016/06/20 18:18:25 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD f7a99e9433) last updated 2016/06/20 18:17:54 (GMT +000)
  config file = 
  configured module search path = Default w/o overrides


```
##### SUMMARY

A different potential fix for https://github.com/ansible/ansible/issues/16370

This is changing apt_repository.py's handling of it's 'mode' argument to be more like the
mode in the common_file_args. 

ie, mode's type='raw', with no default value for the arg, but with a default value in the
python module, set via int(0644, 8)

alikins committed:
Fix default perm for apt_repo files.

A default mode of 0644 (and not specified as int or str)
would get converted to an octal 420, resulting in the
sources file being created with mode '0420' instead of '0644'

Fixes #16370

alikins committed:
Change the file mode arg to 'raw' ala file args

Following the file_common_args model, change the
type of the 'mode' arg here to type='raw' with no
default arg value.

The default mode for file creation is the module
constant DEFAULT_SOURCES_PER, and is used if no
mode os specified.

alikins commented:
ping @abadger

alikins commented:
Tested with:

```
---
- hosts: all
  become: yes
  become_user: root
  tasks:
    - name: delete existing
      file: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list state=absent

    - name: 'no mode specified'
      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present
      register: no_mode_results

    - name: gather no mode stat
      stat: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list
      register: no_mode_stat

    - name: show no mode stat
      debug: var=no_mode_stat

    - name: assert mode is correct
      assert: { that: "no_mode_stat.stat.mode == '0644'" }

    - name: delete 
      file: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list state=absent

    - name: 'mode specified'
      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode=0600
      register: mode_given_results

    - name: gather mode_given stat
      stat: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list
      register: mode_given_stat

    - name: show mode_given_stat
      debug: var=mode_given_stat

    - name: assert mode is correct
      assert: { that: "mode_given_stat.stat.mode == '0600'" }

    - name: delete again
      file: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list state=absent


    - name: 'mode specified as string 0644'
      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode="0600"
      register: mode_given_string_results

    - name: gather mode_given_string stat
      stat: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list
      register: mode_given_string_stat

    - name: show mode_given_string_stat
      debug: var=mode_given_string_stat

    - name: assert mode is correct
      assert: { that: "mode_given_string_stat.stat.mode == '0600'" }

    - name: delete again
      file: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list state=absent

    - name: 'mode specified as string 600'
      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode="600"
      register: mode_given_string_600_results

    - name: gather mode_given_600_string stat
      stat: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list
      register: mode_given_string_600_stat

    - name: show mode_given_string_stat
      debug: var=mode_given_string_600_stat

    - name: assert mode is correct
      assert: { that: "mode_given_string_600_stat.stat.mode == '0600'" }

    - name: delete again
      file: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list state=absent

    - name: 'mode specified as bogus float'
      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode=9.9999999
      register: mode_float_results


    - name: gather mode_floar stat
      stat: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list
      register: mode_float_stat

    - name: show mode_float_stat
      debug: var=mode_float_stat

    - name: delete again for float
      file: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list state=absent


#    - name: 'hey with mode specified'
#      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode=0644

#    - name: 'hey with mode specified as quoted string'
#      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode="0644"

#    - name: 'hey with mode specified as float'
#      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode=1.2345

    - name: 'hey with mode specified as huge int'
      apt_repository: repo='deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main' state=present mode=99999999999999999999999999999999999999999999999999999999999999

```

alikins commented:
Results
(fail expected on the float mode)

```
vagrant@ubuntu:~/src/ansible$ ansible-playbook -i hosts -vvvv apt_repo_test.yml 
No config file found; using defaults
Loaded callback default of type stdout, v2.0

PLAYBOOK: apt_repo_test.yml ****************************************************
1 plays in apt_repo_test.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457405.99-134200814015791 `" && echo ansible-tmp-1466457405.99-134200814015791="` echo $HOME/.ansible/tmp/ansible-tmp-1466457405.99-134200814015791 `" ) && sleep 0'
<localhost> PUT /tmp/tmpN3UcIu TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457405.99-134200814015791/setup
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-jvrnbxapllmmaxnrizuyfoxsqryzsrcm; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457405.99-134200814015791/setup'"'"' && sleep 0'
ok: [localhost]

TASK [delete existing] *********************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:6
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457406.78-104015303995978 `" && echo ansible-tmp-1466457406.78-104015303995978="` echo $HOME/.ansible/tmp/ansible-tmp-1466457406.78-104015303995978 `" ) && sleep 0'
<localhost> PUT /tmp/tmpfFoVf0 TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457406.78-104015303995978/file
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-otimjbehicrgkbbeigiclyhdtlfcysdu; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457406.78-104015303995978/file'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {"after": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}, "before": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "file"}}, "invocation": {"module_args": {"backup": null, "content": null, "delimiter": null, "diff_peek": null, "directory_mode": null, "follow": false, "force": false, "group": null, "mode": null, "original_basename": null, "owner": null, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "recurse": false, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "state": "absent", "validate": null}, "module_name": "file"}, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}

TASK [no mode specified] *******************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:9
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457407.1-244191879376958 `" && echo ansible-tmp-1466457407.1-244191879376958="` echo $HOME/.ansible/tmp/ansible-tmp-1466457407.1-244191879376958 `" ) && sleep 0'
<localhost> PUT /tmp/tmppEYR95 TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457407.1-244191879376958/apt_repository
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-bqbfeycloxwmzmbcbxhcreatviqsjhht; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457407.1-244191879376958/apt_repository'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {}, "invocation": {"module_args": {"filename": null, "install_python_apt": true, "mode": null, "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present", "update_cache": true, "validate_certs": true}, "module_name": "apt_repository"}, "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present"}

TASK [gather no mode stat] *****************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:13
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457419.18-63620266031150 `" && echo ansible-tmp-1466457419.18-63620266031150="` echo $HOME/.ansible/tmp/ansible-tmp-1466457419.18-63620266031150 `" ) && sleep 0'
<localhost> PUT /tmp/tmphXOW0D TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457419.18-63620266031150/stat
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-spndvbgbuveouxnpdlothomlrwhjpnbl; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457419.18-63620266031150/stat'"'"' && sleep 0'
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"checksum_algorithm": "sha1", "follow": false, "get_checksum": true, "get_md5": true, "mime": false, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list"}, "module_name": "stat"}, "stat": {"atime": 1466457407.361988, "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", "ctime": 1466457407.357988, "dev": 64512, "exists": true, "gid": 0, "gr_name": "root", "inode": 4718616, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "md5": "4d446caf1ba7efaaa67d66744d1d4206", "mode": "0644", "mtime": 1466457407.357988, "nlink": 1, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "pw_name": "root", "rgrp": true, "roth": true, "rusr": true, "size": 62, "uid": 0, "wgrp": false, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}}

TASK [show no mode stat] *******************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:17
ok: [localhost] => {
    "no_mode_stat": {
        "changed": false, 
        "stat": {
            "atime": 1466457407.361988, 
            "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", 
            "ctime": 1466457407.357988, 
            "dev": 64512, 
            "exists": true, 
            "gid": 0, 
            "gr_name": "root", 
            "inode": 4718616, 
            "isblk": false, 
            "ischr": false, 
            "isdir": false, 
            "isfifo": false, 
            "isgid": false, 
            "islnk": false, 
            "isreg": true, 
            "issock": false, 
            "isuid": false, 
            "md5": "4d446caf1ba7efaaa67d66744d1d4206", 
            "mode": "0644", 
            "mtime": 1466457407.357988, 
            "nlink": 1, 
            "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", 
            "pw_name": "root", 
            "rgrp": true, 
            "roth": true, 
            "rusr": true, 
            "size": 62, 
            "uid": 0, 
            "wgrp": false, 
            "woth": false, 
            "wusr": true, 
            "xgrp": false, 
            "xoth": false, 
            "xusr": false
        }
    }
}

TASK [assert mode is correct] **************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:20
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"that": "no_mode_stat.stat.mode == '0644'"}, "module_name": "assert"}, "msg": "all assertions passed"}

TASK [delete] ******************************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:23
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457419.4-72103484417590 `" && echo ansible-tmp-1466457419.4-72103484417590="` echo $HOME/.ansible/tmp/ansible-tmp-1466457419.4-72103484417590 `" ) && sleep 0'
<localhost> PUT /tmp/tmpcZm9Ua TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457419.4-72103484417590/file
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-piwrfryuuvkmjboswfvyjgujcflnuqvo; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457419.4-72103484417590/file'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {"after": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}, "before": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "file"}}, "invocation": {"module_args": {"backup": null, "content": null, "delimiter": null, "diff_peek": null, "directory_mode": null, "follow": false, "force": false, "group": null, "mode": null, "original_basename": null, "owner": null, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "recurse": false, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "state": "absent", "validate": null}, "module_name": "file"}, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}

TASK [mode specified] **********************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:26
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457419.54-137846610402756 `" && echo ansible-tmp-1466457419.54-137846610402756="` echo $HOME/.ansible/tmp/ansible-tmp-1466457419.54-137846610402756 `" ) && sleep 0'
<localhost> PUT /tmp/tmpTSFh_T TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457419.54-137846610402756/apt_repository
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-lqkwidjtexthtcfyjuitcttpegxaqeia; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457419.54-137846610402756/apt_repository'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {}, "invocation": {"module_args": {"filename": null, "install_python_apt": true, "mode": "0600", "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present", "update_cache": true, "validate_certs": true}, "module_name": "apt_repository"}, "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present"}

TASK [gather mode_given stat] **************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:30
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457430.62-104372657664369 `" && echo ansible-tmp-1466457430.62-104372657664369="` echo $HOME/.ansible/tmp/ansible-tmp-1466457430.62-104372657664369 `" ) && sleep 0'
<localhost> PUT /tmp/tmplDKK67 TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457430.62-104372657664369/stat
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-qdwgzdhoillsgzihbbgepesousperdqu; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457430.62-104372657664369/stat'"'"' && sleep 0'
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"checksum_algorithm": "sha1", "follow": false, "get_checksum": true, "get_md5": true, "mime": false, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list"}, "module_name": "stat"}, "stat": {"atime": 1466457419.765988, "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", "ctime": 1466457419.761988, "dev": 64512, "exists": true, "gid": 0, "gr_name": "root", "inode": 4720092, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "md5": "4d446caf1ba7efaaa67d66744d1d4206", "mode": "0600", "mtime": 1466457419.761988, "nlink": 1, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "pw_name": "root", "rgrp": false, "roth": false, "rusr": true, "size": 62, "uid": 0, "wgrp": false, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}}

TASK [show mode_given_stat] ****************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:34
ok: [localhost] => {
    "mode_given_stat": {
        "changed": false, 
        "stat": {
            "atime": 1466457419.765988, 
            "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", 
            "ctime": 1466457419.761988, 
            "dev": 64512, 
            "exists": true, 
            "gid": 0, 
            "gr_name": "root", 
            "inode": 4720092, 
            "isblk": false, 
            "ischr": false, 
            "isdir": false, 
            "isfifo": false, 
            "isgid": false, 
            "islnk": false, 
            "isreg": true, 
            "issock": false, 
            "isuid": false, 
            "md5": "4d446caf1ba7efaaa67d66744d1d4206", 
            "mode": "0600", 
            "mtime": 1466457419.761988, 
            "nlink": 1, 
            "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", 
            "pw_name": "root", 
            "rgrp": false, 
            "roth": false, 
            "rusr": true, 
            "size": 62, 
            "uid": 0, 
            "wgrp": false, 
            "woth": false, 
            "wusr": true, 
            "xgrp": false, 
            "xoth": false, 
            "xusr": false
        }
    }
}

TASK [assert mode is correct] **************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:37
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"that": "mode_given_stat.stat.mode == '0600'"}, "module_name": "assert"}, "msg": "all assertions passed"}

TASK [delete again] ************************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:40
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457430.83-90549684724610 `" && echo ansible-tmp-1466457430.83-90549684724610="` echo $HOME/.ansible/tmp/ansible-tmp-1466457430.83-90549684724610 `" ) && sleep 0'
<localhost> PUT /tmp/tmpShF0Nf TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457430.83-90549684724610/file
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-mbcyjufvwcvktpbffahoiqpiwgonqhrt; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457430.83-90549684724610/file'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {"after": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}, "before": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "file"}}, "invocation": {"module_args": {"backup": null, "content": null, "delimiter": null, "diff_peek": null, "directory_mode": null, "follow": false, "force": false, "group": null, "mode": null, "original_basename": null, "owner": null, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "recurse": false, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "state": "absent", "validate": null}, "module_name": "file"}, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}

TASK [mode specified as string 0644] *******************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:44
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457430.96-200191093412320 `" && echo ansible-tmp-1466457430.96-200191093412320="` echo $HOME/.ansible/tmp/ansible-tmp-1466457430.96-200191093412320 `" ) && sleep 0'
<localhost> PUT /tmp/tmpzXlLQR TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457430.96-200191093412320/apt_repository
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-wjfpfecznqcqfjznfexjgrxrezkkjraq; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457430.96-200191093412320/apt_repository'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {}, "invocation": {"module_args": {"filename": null, "install_python_apt": true, "mode": "0600", "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present", "update_cache": true, "validate_certs": true}, "module_name": "apt_repository"}, "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present"}

TASK [gather mode_given_string stat] *******************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:48
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457441.14-272301981825187 `" && echo ansible-tmp-1466457441.14-272301981825187="` echo $HOME/.ansible/tmp/ansible-tmp-1466457441.14-272301981825187 `" ) && sleep 0'
<localhost> PUT /tmp/tmpblBeHF TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457441.14-272301981825187/stat
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-ayvzpdcxnpzflcitzbvmztdnezlvikmj; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457441.14-272301981825187/stat'"'"' && sleep 0'
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"checksum_algorithm": "sha1", "follow": false, "get_checksum": true, "get_md5": true, "mime": false, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list"}, "module_name": "stat"}, "stat": {"atime": 1466457431.173988, "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", "ctime": 1466457431.169988, "dev": 64512, "exists": true, "gid": 0, "gr_name": "root", "inode": 4718616, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "md5": "4d446caf1ba7efaaa67d66744d1d4206", "mode": "0600", "mtime": 1466457431.169988, "nlink": 1, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "pw_name": "root", "rgrp": false, "roth": false, "rusr": true, "size": 62, "uid": 0, "wgrp": false, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}}

TASK [show mode_given_string_stat] *********************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:52
ok: [localhost] => {
    "mode_given_string_stat": {
        "changed": false, 
        "stat": {
            "atime": 1466457431.173988, 
            "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", 
            "ctime": 1466457431.169988, 
            "dev": 64512, 
            "exists": true, 
            "gid": 0, 
            "gr_name": "root", 
            "inode": 4718616, 
            "isblk": false, 
            "ischr": false, 
            "isdir": false, 
            "isfifo": false, 
            "isgid": false, 
            "islnk": false, 
            "isreg": true, 
            "issock": false, 
            "isuid": false, 
            "md5": "4d446caf1ba7efaaa67d66744d1d4206", 
            "mode": "0600", 
            "mtime": 1466457431.169988, 
            "nlink": 1, 
            "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", 
            "pw_name": "root", 
            "rgrp": false, 
            "roth": false, 
            "rusr": true, 
            "size": 62, 
            "uid": 0, 
            "wgrp": false, 
            "woth": false, 
            "wusr": true, 
            "xgrp": false, 
            "xoth": false, 
            "xusr": false
        }
    }
}

TASK [assert mode is correct] **************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:55
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"that": "mode_given_string_stat.stat.mode == '0600'"}, "module_name": "assert"}, "msg": "all assertions passed"}

TASK [delete again] ************************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:58
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457441.34-230625101620550 `" && echo ansible-tmp-1466457441.34-230625101620550="` echo $HOME/.ansible/tmp/ansible-tmp-1466457441.34-230625101620550 `" ) && sleep 0'
<localhost> PUT /tmp/tmpofKEFw TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457441.34-230625101620550/file
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-hkeygkhffjsckewqrxfhpchfvgnugojz; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457441.34-230625101620550/file'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {"after": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}, "before": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "file"}}, "invocation": {"module_args": {"backup": null, "content": null, "delimiter": null, "diff_peek": null, "directory_mode": null, "follow": false, "force": false, "group": null, "mode": null, "original_basename": null, "owner": null, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "recurse": false, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "state": "absent", "validate": null}, "module_name": "file"}, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}

TASK [mode specified as string 600] ********************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:61
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457441.47-99258021514168 `" && echo ansible-tmp-1466457441.47-99258021514168="` echo $HOME/.ansible/tmp/ansible-tmp-1466457441.47-99258021514168 `" ) && sleep 0'
<localhost> PUT /tmp/tmpEXeaoB TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457441.47-99258021514168/apt_repository
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-sgtuesnpfplapgryfumnqzrkoiiroqfk; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457441.47-99258021514168/apt_repository'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {}, "invocation": {"module_args": {"filename": null, "install_python_apt": true, "mode": "600", "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present", "update_cache": true, "validate_certs": true}, "module_name": "apt_repository"}, "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present"}

TASK [gather mode_given_600_string stat] ***************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:65
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457453.16-179999928672399 `" && echo ansible-tmp-1466457453.16-179999928672399="` echo $HOME/.ansible/tmp/ansible-tmp-1466457453.16-179999928672399 `" ) && sleep 0'
<localhost> PUT /tmp/tmpiW0Bj8 TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457453.16-179999928672399/stat
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-zzriwxppvdglknnrvlnobatcquuitvmc; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457453.16-179999928672399/stat'"'"' && sleep 0'
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"checksum_algorithm": "sha1", "follow": false, "get_checksum": true, "get_md5": true, "mime": false, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list"}, "module_name": "stat"}, "stat": {"atime": 1466457446.793988, "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", "ctime": 1466457441.685988, "dev": 64512, "exists": true, "gid": 0, "gr_name": "root", "inode": 4720092, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "md5": "4d446caf1ba7efaaa67d66744d1d4206", "mode": "0600", "mtime": 1466457441.685988, "nlink": 1, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "pw_name": "root", "rgrp": false, "roth": false, "rusr": true, "size": 62, "uid": 0, "wgrp": false, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}}

TASK [show mode_given_string_stat] *********************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:69
ok: [localhost] => {
    "mode_given_string_600_stat": {
        "changed": false, 
        "stat": {
            "atime": 1466457446.793988, 
            "checksum": "cab76bc2661eb55883885bddc35483afb1dc433e", 
            "ctime": 1466457441.685988, 
            "dev": 64512, 
            "exists": true, 
            "gid": 0, 
            "gr_name": "root", 
            "inode": 4720092, 
            "isblk": false, 
            "ischr": false, 
            "isdir": false, 
            "isfifo": false, 
            "isgid": false, 
            "islnk": false, 
            "isreg": true, 
            "issock": false, 
            "isuid": false, 
            "md5": "4d446caf1ba7efaaa67d66744d1d4206", 
            "mode": "0600", 
            "mtime": 1466457441.685988, 
            "nlink": 1, 
            "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", 
            "pw_name": "root", 
            "rgrp": false, 
            "roth": false, 
            "rusr": true, 
            "size": 62, 
            "uid": 0, 
            "wgrp": false, 
            "woth": false, 
            "wusr": true, 
            "xgrp": false, 
            "xoth": false, 
            "xusr": false
        }
    }
}

TASK [assert mode is correct] **************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:72
ok: [localhost] => {"changed": false, "invocation": {"module_args": {"that": "mode_given_string_600_stat.stat.mode == '0600'"}, "module_name": "assert"}, "msg": "all assertions passed"}

TASK [delete again] ************************************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:75
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457453.39-138050600289821 `" && echo ansible-tmp-1466457453.39-138050600289821="` echo $HOME/.ansible/tmp/ansible-tmp-1466457453.39-138050600289821 `" ) && sleep 0'
<localhost> PUT /tmp/tmpYnMI3l TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457453.39-138050600289821/file
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-vqqjhfysubgbedillbwanrdjtktzqmjv; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457453.39-138050600289821/file'"'"' && sleep 0'
changed: [localhost] => {"changed": true, "diff": {"after": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}, "before": {"path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "file"}}, "invocation": {"module_args": {"backup": null, "content": null, "delimiter": null, "diff_peek": null, "directory_mode": null, "follow": false, "force": false, "group": null, "mode": null, "original_basename": null, "owner": null, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "recurse": false, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "state": "absent", "validate": null}, "module_name": "file"}, "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "state": "absent"}

TASK [mode specified as bogus float] *******************************************
task path: /home/vagrant/src/ansible/apt_repo_test.yml:78
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466457453.53-59266786871114 `" && echo ansible-tmp-1466457453.53-59266786871114="` echo $HOME/.ansible/tmp/ansible-tmp-1466457453.53-59266786871114 `" ) && sleep 0'
<localhost> PUT /tmp/tmplORDJQ TO /home/vagrant/.ansible/tmp/ansible-tmp-1466457453.53-59266786871114/apt_repository
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-vxdestkisqymfhsftwwqerlkhxtunjxo; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1466457453.53-59266786871114/apt_repository'"'"' && sleep 0'
fatal: [localhost]: FAILED! => {"changed": false, "details": "bad symbolic permission for mode: 9.9999999", "failed": true, "gid": 0, "group": "root", "invocation": {"module_args": {"filename": null, "install_python_apt": true, "mode": "9.9999999", "repo": "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main", "state": "present", "update_cache": true, "validate_certs": true}, "module_name": "apt_repository"}, "mode": "0644", "msg": "mode must be in octal or symbolic form", "owner": "root", "path": "/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list", "size": 0, "state": "file", "uid": 0}

NO MORE HOSTS LEFT *************************************************************
 [WARNING]: Could not create retry file 'apt_repo_test.retry'.         [Errno 2] No such file or directory: ''


PLAY RECAP *********************************************************************
localhost                  : ok=22   changed=9    unreachable=0    failed=1   
```

gregdek commented:
Thanks @alikins. All backport requests must be reviewed by the core team, and this can take time. We appreciate your patience. 

[This message brought to you by your friendly Ansibull-bot.]

alikins commented:
updated the test playbook to use yaml dict style task args, and added some tests.
playbook:
https://gist.github.com/alikins/8d3c44f44e77888acb7585de5e483373

results: https://gist.github.com/alikins/c7b16f408fa7cae810e149931a037bb6

The test that fails is:

```
    - name: 'mode specified short'
      apt_repository: 
        repo: 'deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main'
        state: present
        mode: 600
      register: mode_given_short_results

    - name: gather mode_given stat
      stat: path=/etc/apt/sources.list.d/apt_postgresql_org_pub_repos_apt.list
      register: mode_given_short_stat

    - name: show mode_given_short_stat
      debug: var=mode_given_short_stat

    - name: assert mode is correct
      assert: { that: "mode_given_short_stat.stat.mode == '0600'" }
```

Where the resultant mode ends up             "mode": "1130", 

I'm not entirely sure if thats expected behavior or not.

abadger commented:
@alikins yeah, this test "should" fail.  yaml has no way to tell that 644 in that play is supposed to be an octal number so it interprets it as a base-10 decimal and comes out with something unexpected.  There's nothing we can do about that except mention it in the documentation.  "600" should work.  As well as 0600.  But 600 will yield something unexpected.

alikins commented:
Depending on https://github.com/ansible/ansible-modules-core/issues/4072, if it gets merged, we can cherry pick the commit for stable-2.1 and close this pr.

gregdek commented:
Thanks @alikins for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

alikins commented:
cherrypicked to stable-2.1 in da177b17449e519a77a9971f18b93eac58aee065
